crates/matrix-sdk/src/client/mod.rs

```diff
@@ -1560,12 +1568,47 @@ impl Client {
     /// # Arguments
     ///
     /// * `room_id` - The `RoomId` of the room to be joined.
-    pub async fn join_room_by_id(
-        &self,
-        room_id: &RoomId,
-    ) -> HttpResult<join_room_by_id::v3::Response> {
+    pub async fn join_room_by_id(&self, room_id: &RoomId) -> Result<room::Joined> {
         let request = join_room_by_id::v3::Request::new(room_id);
-        self.send(request, None).await
+
+        let (tx, mut rx) = mpsc::channel::<Result<room::Joined>>(1);
+
+        let handle = self.add_room_event_handler(room_id, {
+            move |event: SyncStateEvent<RoomMemberEventContent>,
+                  room: room::Room,
+                  client: Client| {
+                let tx = tx.clone();
+
+                async move {
+                    if event.membership() == &MembershipState::Join
+                        && event.state_key() == client.user_id().expect("user_id")
```

Comment:
How about extracting the user ID before the adding the event handler and bubbling up that error, instead of panicking here?

Reply:
Good suggestion, the question as which Error it should be bubbled up remains

Reply:
I've used `InconsistentState` to begin with, although I am reasonably unhappy with it. Calling a method that requires being logged is a more general issue and should probably throw it's own error, wouldn't you agree?

Reply:
Found this in another part of the code, seems fitting:

```rust
let user_id = self.user_id().ok_or(Error::AuthenticationRequired)?;
```